recursion/src/air/alu/air.rs

```diff
@@ -0,0 +1,464 @@
+    where
+        StandardUniform: Distribution<F>,
+    {
+        let mut rng = SmallRng::seed_from_u64(1);
+        let n_padded = rows.next_power_of_two();
+        let mut trace =
+            RowMajorMatrix::new(F::zero_vec(n_padded * Self::TRACE_WIDTH), Self::TRACE_WIDTH);
+
+        let (prefix, rows, suffix) = unsafe { trace.values.align_to_mut::<AluCols<F, R>>() };
+        assert!(prefix.is_empty(), "Alignment should match");
+        assert!(suffix.is_empty(), "Alignment should match");
+        assert_eq!(rows.len(), n_padded);
+
+        for (i, row) in rows.iter_mut().enumerate() {
+            for (left_addr, left_val, right_addr, right_val, res_addr, res_val) in izip!(
+                row.left_addr.iter_mut(),
+                row.left_val.iter_mut(),
+                row.right_addr.iter_mut(),
+                row.right_val.iter_mut(),
+                row.res_addr.iter_mut(),
+                row.res_val.iter_mut(),
+            ) {
+                *left_addr = F::from_usize(3 * i);
+                *left_val = rng.random();
+                *right_addr = F::from_usize(3 * i + 1);
+                *right_val = rng.random();
+                *res_addr = F::from_usize(3 * i + 2);
+                *res_val = self.apply(*left_val, *right_val);
+                if !valid {
+                    // make it invalid
+                    *res_val *= F::TWO;
+                }
+            }
+        }
+        trace
+    }
+
+    fn width(&self) -> usize {
+        Self::TRACE_WIDTH
+    }
+
+    fn eval(&self, builder: &mut AB) {
+        let main = builder.main();
+        let local = main.row_slice(0).expect("Matrix is empty?");
+        let local: &AluCols<AB::Var> = (*local).borrow();
+
+        for ((left_val, right_val), res_val) in local
+            .left_val
+            .iter()
+            .zip(local.right_val.iter())
+            .zip(local.res_val.iter())
+        {
+            builder.assert_eq(
+                self.apply(left_val.clone(), right_val.clone()),
+                res_val.clone(),
+            );
+        }
+    }
+}
+
+#[macro_export]
+macro_rules! operation_air {
+    ($OpAir:ident, $op:tt, $events:ident) => {
+        pub struct $OpAir<const REPETITIONS: usize = 1> {
+        }
+
+        impl<const REPETITIONS: usize> $OpAir<REPETITIONS> {
+            pub const fn new() -> Self {
+                Self {  }
+            }
+        }
+
+        impl<const REPETITIONS: usize> GenerateAluTrace<REPETITIONS> for $OpAir<REPETITIONS> {}
+
+        // your per-op behavior
+        impl<const REPETITIONS: usize> $crate::air::alu::air::Operation for $OpAir<REPETITIONS> {
+            fn apply<I, O>(&self, left: I, right: I) -> O
+            where
+                I: core::ops::Add<Output = O>
+                 + core::ops::Sub<Output = O>
+                 + core::ops::Mul<Output = O>,
+            { left $op right }
+        }
```

Comment:
This is a bit weird, because we're adding bounds to support the different operations Add / Sub / Mul, but what if we add more down the road? Wouldn't it be best to make it generic over some single `Op` bound?